File: src/BigPictureTV.py
import os
import sys
import json
from PyQt6.QtWidgets import QApplication, QSystemTrayIcon, QMenu, QMainWindow, QMessageBox
from PyQt6.QtGui import QIcon, QAction
from PyQt6.QtCore import Qt, QTimer, QSharedMemory, QTranslator, QLocale
from settings_window import Ui_SettingsWindow
from tooltiped_slider import TooltipedSlider
import audio_manager as AudioManager
import mode_manager as ModeManager
import shortcut_manager as ShortcutManager
import color_utils as ColorUtils
import utils

SETTINGS_FILE = os.path.join(os.environ["APPDATA"], "BigPictureTV", "settings.json")
ICONS_FOLDER = "icons" if getattr(sys, "frozen", False) else os.path.join(os.path.dirname(__file__), "icons")


class BigPictureTV(QMainWindow):
    def __init__(self):
        super().__init__()
        self.ui = Ui_SettingsWindow()
        self.ui.setupUi(self)
        self.setWindowIcon(QIcon(os.path.join(ICONS_FOLDER, f"icon_desktop_{utils.get_theme()}.png")))
        self.current_mode = ModeManager.read_current_mode()
        self.settings = {}
        self.first_run = False
        self.paused = False
        self.set_fusion_frames()
        self.init_checkrate_tooltipedslider()
        self.populate_comboboxes()
        self.load_settings()
        self.setup_ui_connections()
        self.get_audio_capabilities()
        self.tray_icon = self.create_tray_icon()
        self.timer = QTimer()
        self.timer.timeout.connect(self.update_mode)
        self.update_mode_timer_interval(self.ui.checkrate_slider.value())
        if self.first_run:
            self.show()
            self.first_run = False

    def set_fusion_frames(self):
        if app.style().objectName() == "fusion":
            ColorUtils.set_frame_color_based_on_window(self, self.ui.actions_frame)
            ColorUtils.set_frame_color_based_on_window(self, self.ui.monitor_frame)
            ColorUtils.set_frame_color_based_on_window(self, self.ui.audio_frame)
            ColorUtils.set_frame_color_based_on_window(self, self.ui.settings_frame)

    def setup_ui_connections(self):
        self.ui.disable_audio_checkbox.stateChanged.connect(self.handle_disableaudio_checkbox_state_changed)
        self.ui.gamemode_audio_lineedit.textChanged.connect(self.save_settings)
        self.ui.desktop_audio_lineedit.textChanged.connect(self.save_settings)
        self.ui.startup_checkbox.setChecked(ShortcutManager.check_startup_shortcut())
        self.ui.close_discord_checkbox.stateChanged.connect(self.save_settings)
        self.ui.close_discord_checkbox.setEnabled(utils.is_discord_installed())
        self.ui.close_discord_label.setEnabled(utils.is_discord_installed())
        self.ui.startup_checkbox.stateChanged.connect(self.handle_startup_checkbox_state_changed)
        self.ui.gamemode_monitor_combobox.currentIndexChanged.connect(self.save_settings)
        self.ui.desktop_monitor_combobox.currentIndexChanged.connect(self.save_settings)
        self.ui.checkrate_slider.valueChanged.connect(self.handle_checkrate_slider_value_changed)
        self.ui.checkrate_slider.sliderReleased.connect(self.save_settings)
        self.ui.install_audio_button.clicked.connect(self.handle_audio_button_clicked)
        self.ui.disable_monitor_checkbox.stateChanged.connect(self.handle_disablemonitor_checkbox_state_changed)

    def populate_comboboxes(self):
        self.ui.gamemode_monitor_combobox.addItem(self.tr("External"))
        self.ui.gamemode_monitor_combobox.addItem(self.tr("Clone"))
        self.ui.desktop_monitor_combobox.addItem(self.tr("Internal"))
        self.ui.desktop_monitor_combobox.addItem(self.tr("Extend"))

    def init_checkrate_tooltipedslider(self):
        self.checkrate_tooltipedslider = TooltipedSlider(Qt.Orientation.Horizontal, self.ui.centralwidget)
        layout = self.ui.settings_layout
        index = layout.indexOf(self.ui.checkrate_label)
        position = layout.getItemPosition(index)
        row = position[0]
        column = position[1]
        layout.addWidget(self.checkrate_tooltipedslider, row, column + 1)
        self.ui.checkrate_slider.deleteLater()
        self.ui.checkrate_slider = self.checkrate_tooltipedslider
        self.ui.checkrate_slider.setTickInterval(100)
        self.ui.checkrate_slider.setRange(100, 1000)

    def handle_startup_checkbox_state_changed(self, state):
        ShortcutManager.manage_startup_shortcut(state)

    def handle_audio_button_clicked(self):
        self.ui.install_audio_button.setEnabled(False)
        status, message = utils.install_audio_module()
        if status == "Success":
            QMessageBox.information(self, status, message)
        else:
            QMessageBox.critical(self, status, message)
            self.ui.install_audio_button.setEnabled(True)

        self.get_audio_capabilities()

    def handle_checkrate_slider_value_changed(self, value):
        remainder = value % 100
        if remainder != 0:
            if remainder < 50:
                value -= remainder
            else:
                value += 100 - remainder
            self.ui.checkrate_slider.setValue(value)

        self.update_mode_timer_interval(value)

    def handle_disableaudio_checkbox_state_changed(self, state):
        self.toggle_audio_settings(not state)
        self.save_settings()

    def handle_disablemonitor_checkbox_state_changed(self, state):
        self.toggle_monitor_settings(not state)
        self.save_settings()

    def toggle_audio_settings(self, enabled):
        self.ui.desktop_audio_lineedit.setEnabled(enabled)
        self.ui.desktop_audio_label.setEnabled(enabled)
        self.ui.gamemode_audio_lineedit.setEnabled(enabled)
        self.ui.gamemode_audio_label.setEnabled(enabled)
        self.ui.audio_output_label.setEnabled(enabled)

    def toggle_monitor_settings(self, enabled):
        self.ui.gamemode_monitor_combobox.setEnabled(enabled)
        self.ui.desktop_monitor_combobox.setEnabled(enabled)
        self.ui.gamemode_monitor_label.setEnabled(enabled)
        self.ui.desktop_monitor_label.setEnabled(enabled)
        self.ui.monitor_configuration_label.setEnabled(enabled)

    def create_default_settings(self):
        self.settings = {
            "gamemode_audio": "TV",
            "desktop_audio": "Headset",
            "disable_audio_switch": False,
            "checkrate": 1000,
        }
        self.apply_settings()
        self.save_settings()
        self.first_run = True

    def load_settings(self):
        if not os.path.exists(os.path.dirname(SETTINGS_FILE)):
            os.makedirs(os.path.dirname(SETTINGS_FILE))
        if not os.path.exists(SETTINGS_FILE):
            self.create_default_settings()
        else:
            with open(SETTINGS_FILE, "r") as f:
                self.settings = json.load(f)
        self.apply_settings()

    def apply_settings(self):
        self.ui.gamemode_audio_lineedit.setText(self.settings.get("gamemode_audio", ""))
        self.ui.desktop_audio_lineedit.setText(self.settings.get("desktop_audio", ""))
        self.ui.disable_audio_checkbox.setChecked(self.settings.get("disable_audio_switch", False))
        self.ui.checkrate_slider.setValue(self.settings.get("checkrate", 1000))
        self.ui.close_discord_checkbox.setChecked(self.settings.get("discord_action", False))
        self.ui.gamemode_monitor_combobox.setCurrentIndex(self.settings.get("gamemode_monitor", 0))
        self.ui.desktop_monitor_combobox.setCurrentIndex(self.settings.get("desktop_monitor", 0))
        self.ui.disable_monitor_checkbox.setChecked(self.settings.get("disable_monitor_switch", False))
        self.toggle_audio_settings(not self.ui.disable_audio_checkbox.isChecked())
        self.toggle_monitor_settings(not self.ui.disable_monitor_checkbox.isChecked())

    def save_settings(self):
        self.settings = {
            "gamemode_audio": self.ui.gamemode_audio_lineedit.text(),
            "desktop_audio": self.ui.desktop_audio_lineedit.text(),
            "disable_audio_switch": self.ui.disable_audio_checkbox.isChecked(),
            "checkrate": self.ui.checkrate_slider.value(),
            "discord_action": self.ui.close_discord_checkbox.isChecked(),
            "gamemode_monitor": self.ui.gamemode_monitor_combobox.currentIndex(),
            "desktop_monitor": self.ui.desktop_monitor_combobox.currentIndex(),
            "disable_monitor_switch": self.ui.disable_monitor_checkbox.isChecked(),
        }
        os.makedirs(os.path.dirname(SETTINGS_FILE), exist_ok=True)
        with open(SETTINGS_FILE, "w") as f:
            json.dump(self.settings, f, indent=4)

    def switch_mode(self, mode):
        if mode == self.current_mode:
            return

        self.current_mode = mode
        self.handle_monitor_changes(self.current_mode)
        self.handle_actions(self.current_mode)
        self.handle_audio_changes(self.current_mode)
        ModeManager.write_current_mode(self.current_mode)

        if self.tray_icon:
            variant = "gamemode" if self.current_mode == ModeManager.Mode.GAMEMODE else "desktop"
            self.tray_icon.setIcon(QIcon(os.path.join(ICONS_FOLDER, f"icon_{variant}_{utils.get_theme()}.png")))
            self.tray_icon.setContextMenu(self.create_menu())

    def handle_monitor_changes(self, mode):
        if not self.ui.disable_monitor_checkbox.isChecked():
            if mode == ModeManager.Mode.GAMEMODE and self.ui.gamemode_monitor_combobox.currentIndex() == 0:
                monitor_mode = "/external"
            elif mode == ModeManager.Mode.GAMEMODE and self.ui.gamemode_monitor_combobox.currentIndex() == 1:
                monitor_mode = "/clone"
            elif mode == ModeManager.Mode.DESKTOP and self.ui.desktop_monitor_combobox.currentIndex() == 0:
                monitor_mode = "/internal"
            elif mode == ModeManager.Mode.DESKTOP and self.ui.desktop_monitor_combobox.currentIndex() == 1:
                monitor_mode = "/extend"
            utils.run_displayswitch(monitor_mode)

    def handle_actions(self, mode):
        if self.ui.close_discord_checkbox.isChecked():
            utils.close_discord() if mode == ModeManager.Mode.GAMEMODE else utils.start_discord()

    def handle_audio_changes(self, mode):
        if not self.ui.disable_audio_checkbox.isChecked():
            gamemode_audio = self.ui.gamemode_audio_lineedit.text()
            desktop_audio = self.ui.desktop_audio_lineedit.text()
            AudioManager.switch_audio(gamemode_audio if mode == ModeManager.Mode.GAMEMODE else desktop_audio)

    def get_audio_capabilities(self):
        if not utils.is_audio_device_cmdlets_installed():
            self.ui.disable_audio_checkbox.setChecked(True)
            self.ui.disable_audio_checkbox.setEnabled(False)
            self.toggle_audio_settings(False)
        else:
            self.ui.disable_audio_checkbox.setEnabled(True)
            self.ui.install_audio_button.setVisible(False)
            if not self.ui.disable_audio_checkbox.isChecked():
                self.toggle_audio_settings(True)
            self.adjustSize()

    def update_mode(self):
        if (
            utils.is_bigpicture_running()
            and self.current_mode != ModeManager.Mode.GAMEMODE
            and not utils.is_sunshine_stream_active()
        ):
            self.switch_mode(ModeManager.Mode.GAMEMODE)
        elif not utils.is_bigpicture_running() and self.current_mode != ModeManager.Mode.DESKTOP:
            self.switch_mode(ModeManager.Mode.DESKTOP)

    def update_mode_timer_interval(self, check_rate):
        self.timer.setInterval(check_rate)
        self.timer.start()

    def create_tray_icon(self):
        tray_icon = QSystemTrayIcon(QIcon(os.path.join(ICONS_FOLDER, f"icon_desktop_{utils.get_theme()}.png")))
        tray_icon.setToolTip("BigPictureTV")
        tray_icon.setContextMenu(self.create_menu())
        tray_icon.show()
        return tray_icon

    def create_menu(self):
        menu = QMenu()
        pause_resume_action = QAction(self.tr("Resume detection") if self.paused else self.tr("Pause detection"), menu)
        pause_resume_action.triggered.connect(self.toggle_detection)
        settings_action = QAction(self.tr("Settings"), menu)
        settings_action.triggered.connect(self.show)
        exit_action = QAction(self.tr("Exit"), menu)
        exit_action.triggered.connect(QApplication.quit)

        menu.addAction(pause_resume_action)
        menu.addAction(settings_action)
        menu.addAction(exit_action)

        return menu

    def toggle_detection(self):
        self.paused = not self.paused
        if self.paused:
            self.timer.stop()
            self.tray_icon.setToolTip("BigPictureTV (Paused)")
        else:
            self.timer.start()
            self.tray_icon.setToolTip("BigPictureTV")
        self.tray_icon.setContextMenu(self.create_menu())

    def closeEvent(self, event):
        event.ignore()
        self.hide()


if __name__ == "__main__":
    shared_memory = QSharedMemory("BigPictureTVSharedMemory")
    if shared_memory.attach() or not shared_memory.create(1):
        sys.exit(0)

    app = QApplication(sys.argv)
    if utils.is_windows_10():
        app.setStyle("Fusion")

    translator = QTranslator()
    locale_name = QLocale.system().name()
    locale = locale_name[:2]
    if locale:
        file_name = f"tr/bigpicturetv_{locale}.qm"
    else:
        file_name = None

    if file_name and translator.load(file_name):
        app.installTranslator(translator)

    big_picture_tv = BigPictureTV()
    sys.exit(app.exec())

File: src/bigpicturetv.cpp
#include "bigpicturetv.h"
#include "colorutils.h"
#include "shortcutmanager.h"
#include "steamwindowmanager.h"
#include "ui_bigpicturetv.h"
#include "audiomanager.h"
#include "utils.h"

#include <QDir>
#include <QStandardPaths>
#include <QFile>
#include <QIcon>
#include <QProcess>
#include <QJsonDocument>
#include <QJsonParseError>
#include <QSystemTrayIcon>
#include <QMenu>
#include <QAction>
#include <QApplication>
#include <QCloseEvent>
#include <QMessageBox>
#include <QDebug>
#include <iostream>

const QString BigPictureTV::settingsFile = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation) + "/BigPictureTV/settings.json";

BigPictureTV::BigPictureTV(QWidget *parent)
    : QMainWindow(parent)
    , gamemodeActive(false)
    , firstRun(false)
    , ui(new Ui::BigPictureTV)
    , windowCheckTimer(new QTimer(this))

{
    ui->setupUi(this);
    setWindowIcon(getIconForTheme());
    setFrames();
    populateComboboxes();
    loadSettings();
    setupConnections();
    getAudioCapabilities();


    windowCheckTimer->setInterval(ui->checkrate_slider->value());
    connect(windowCheckTimer, &QTimer::timeout, this, &BigPictureTV::checkWindowTitle);
    windowCheckTimer->start();
    createTrayIcon();
    if (firstRun)
    {
        this->show();
    }
}

BigPictureTV::~BigPictureTV()
{
    saveSettings();
    delete ui;
}

void BigPictureTV::setupConnections()
{
    connect(ui->startup_checkbox, &QCheckBox::stateChanged, this, &BigPictureTV::onStartupCheckboxStateChanged);
    connect(ui->desktop_audio_lineedit, &QLineEdit::textChanged, this, &BigPictureTV::saveSettings);
    connect(ui->gamemode_audio_lineedit, &QLineEdit::textChanged, this, &BigPictureTV::saveSettings);
    connect(ui->disable_audio_checkbox, &QCheckBox::stateChanged, this, &BigPictureTV::onDisableAudioCheckboxStateChanged);
    connect(ui->disable_monitor_checkbox, &QCheckBox::stateChanged, this, &BigPictureTV::onDisableMonitorCheckboxStateChanged);
    connect(ui->checkrate_slider, &QSlider::sliderReleased, this, &BigPictureTV::onCheckrateSliderReleased);
    ui->close_discord_checkbox->setEnabled(isDiscordInstalled());
    ui->close_discord_label->setEnabled(isDiscordInstalled());
    connect(ui->close_discord_checkbox, &QCheckBox::stateChanged, this, &BigPictureTV::saveSettings);
    connect(ui->performance_powerplan_checkbox, &QCheckBox::stateChanged, this, &BigPictureTV::saveSettings);
    connect(ui->desktop_monitor_combobox, &QComboBox::currentIndexChanged, this, &BigPictureTV::saveSettings);
    connect(ui->gamemode_monitor_combobox, &QComboBox::currentIndexChanged, this, &BigPictureTV::saveSettings);
    connect(ui->install_audio_button, &QPushButton::clicked, this, &BigPictureTV::onAudioButtonClicked);
}

void BigPictureTV::getAudioCapabilities()
{
    if (!isAudioDeviceCmdletsInstalled())
    {
        qDebug() << "here:";
        ui->disable_audio_checkbox->setChecked(true);
        ui->disable_audio_checkbox->setEnabled(false);
        toggleAudioSettings(false);
    }
    else
    {
        qDebug() << "there:";
        ui->disable_audio_checkbox->setEnabled(true);
        ui->install_audio_button->setVisible(false);
        if (!ui->disable_audio_checkbox->isChecked())
        {
            toggleAudioSettings(true);
        }
        this->adjustSize();
    }
}

void BigPictureTV::setFrames()
{
    setFrameColorBasedOnWindow(this, ui->actions_frame);
    setFrameColorBasedOnWindow(this, ui->audio_frame);
    setFrameColorBasedOnWindow(this, ui->monitor_frame);
    setFrameColorBasedOnWindow(this, ui->settings_frame);
}

void BigPictureTV::populateComboboxes()
{
    ui->desktop_monitor_combobox->addItem("Internal");
    ui->desktop_monitor_combobox->addItem("Extend");
    ui->gamemode_monitor_combobox->addItem("External");
    ui->gamemode_monitor_combobox->addItem("Clone");
}

void BigPictureTV::createTrayIcon()
{
    trayIcon = new QSystemTrayIcon(getIconForTheme(), this);
    trayIcon->setToolTip("BigPictureTV");
    trayIcon->setContextMenu(createMenu());
    trayIcon->show();
}

QMenu* BigPictureTV::createMenu()
{
    QMenu *menu = new QMenu(this);

    QAction *settingsAction = new QAction("Settings", this);
    connect(settingsAction, &QAction::triggered, this, &BigPictureTV::showSettings);

    QAction *exitAction = new QAction("Exit", this);
    connect(exitAction, &QAction::triggered, QApplication::instance(), &QApplication::quit);

    menu->addAction(settingsAction);
    menu->addAction(exitAction);

    return menu;
}

void BigPictureTV::closeEvent(QCloseEvent *event)
{
    hide();
    event->ignore();
}

void BigPictureTV::showSettings()
{
    this->show();
}

void BigPictureTV::onCheckrateSliderReleased()
{
    int newInterval = ui->checkrate_slider->value();
    windowCheckTimer->setInterval(newInterval);
    saveSettings();
    qDebug() << "Checkrate updated to:" << newInterval << "ms";
}


void BigPictureTV::onStartupCheckboxStateChanged()
{
    manageShortcut(shortcutName, ui->startup_checkbox->isChecked());
}

void BigPictureTV::onDisableAudioCheckboxStateChanged(int state)
{
    bool isChecked = (state == Qt::Checked);
    toggleAudioSettings(!isChecked);
    saveSettings();
}

void BigPictureTV::onDisableMonitorCheckboxStateChanged(int state)
{
    bool isChecked = (state == Qt::Checked);
    toggleMonitorSettings(!isChecked);
    saveSettings();
}

void BigPictureTV::onAudioButtonClicked()
{
    ui->install_audio_button->setEnabled(false);

    QProcess process;
    process.start("powershell", QStringList() << "-NoProfile" << "-ExecutionPolicy" << "Bypass" << "-Command"
                                              << "Install-Module AudioDeviceCmdlets -Force -Scope CurrentUser");

    if (!process.waitForFinished()) {
        status = "Error";
        message = "Failed to execute the PowerShell command.\n"
                  "Please check if PowerShell is installed and properly configured.";
    } else {
        QString output = process.readAllStandardOutput();
        QString errorOutput = process.readAllStandardError();
        int exitCode = process.exitCode();

        if (exitCode == 0) {
            status = "Success";
            message = "AudioDeviceCmdlets module installed successfully.\nYou can now use audio settings.";
        } else {
            status = "Error";
            message = "Failed to install AudioDeviceCmdlets module.\n"
                      "Please install it manually by running this command in PowerShell: "
                      "Install-Module AudioDeviceCmdlets -Force -Scope CurrentUser\n"
                      "You should then restart the application.\n"
                      "Error details:\n" + errorOutput;
        }
    }

    if (status == "Success") {
        QMessageBox::information(this, status, message);
    } else {
        QMessageBox::critical(this, status, message);
        ui->install_audio_button->setEnabled(true);
    }
    getAudioCapabilities();
}


void BigPictureTV::checkWindowTitle()
{
    //qDebug() << "checking";
    bool isRunning = isBigPictureRunning();
    bool disableVideo = ui->disable_monitor_checkbox->isChecked();
    bool disableAudio = ui->disable_audio_checkbox->isChecked();

    if (isRunning && !gamemodeActive)
    {
        gamemodeActive = true;
        handleMonitorChanges(false, disableVideo);
        handleAudioChanges(false, disableAudio);
        handleActions(false);
    }
    else if (!isRunning && gamemodeActive)
    {
        gamemodeActive = false;
        handleMonitorChanges(true, disableVideo);
        handleAudioChanges(true, disableAudio);
        handleActions(true);
    }
}

void BigPictureTV::handleMonitorChanges(bool isDesktopMode, bool disableVideo)
{
    if (disableVideo) return;

    int index = isDesktopMode ? ui->desktop_monitor_combobox->currentIndex() : ui->gamemode_monitor_combobox->currentIndex();

    if (isDesktopMode)
    {
        if (index == 0)
        {
            runDisplayswitch("/internal");
        }
        else if (index == 1)
        {
            runDisplayswitch("/extend");
        }
    }
    else
    {
        if (index == 0)
        {
            runDisplayswitch("/external");
        }
        else if (index == 1)
        {
            runDisplayswitch("/clone");
        }
    }
}

void BigPictureTV::handleAudioChanges(bool isDesktopMode, bool disableAudio)
{
    if (disableAudio) return;

    std::string audioDevice = isDesktopMode ? ui->desktop_audio_lineedit->text().toStdString() : ui->gamemode_audio_lineedit->text().toStdString();

    try {
        setAudioDevice(audioDevice);
        std::cout << "Audio device set successfully." << std::endl;
    } catch (const std::runtime_error& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }
}

void BigPictureTV::handleActions(bool isDesktopMode)
{
    bool performDiscordAction = ui->close_discord_checkbox->isChecked();
    bool setPowerPlan = ui->performance_powerplan_checkbox->isChecked();
    if (isDesktopMode)
    {
        if (performDiscordAction)
        {
            startDiscord();
        }
        if (setPowerPlan)
        {
            switchPowerPlan(L"381b4222-f694-41f0-9685-ff5bb260df2e");
        }
    }
    else
    {
        if (performDiscordAction)
        {
            closeDiscord();
        }
        if (setPowerPlan)
        {
            switchPowerPlan(L"8c5e7fda-e8bf-4a96-9a85-a6e23a8c635c");
        }
    }

}

void BigPictureTV::createDefaultSettings()
{
    firstRun = true;
    ui->desktop_audio_lineedit->setText("Headset");
    ui->gamemode_audio_lineedit->setText("TV");
    saveSettings();
}

void BigPictureTV::loadSettings()
{
    QDir settingsDir(QFileInfo(settingsFile).absolutePath());
    if (!settingsDir.exists()) {
        settingsDir.mkpath(settingsDir.absolutePath());
    }

    QFile file(settingsFile);
    if (!file.exists()) {
        createDefaultSettings();

    } else {
        if (file.open(QIODevice::ReadOnly)) {
            QJsonParseError parseError;
            QJsonDocument doc = QJsonDocument::fromJson(file.readAll(), &parseError);
            if (parseError.error == QJsonParseError::NoError) {
                settings = doc.object();
            } else {
                qWarning() << "Error parsing settings file:" << parseError.errorString();
            }
            file.close();
        } else {
            qWarning() << "Failed to open settings file for reading.";
        }
    }
    applySettings();
}

void BigPictureTV::applySettings()
{
    qDebug() << "applying settings.";
    ui->gamemode_audio_lineedit->setText(settings.value("gamemode_audio").toString());
    ui->desktop_audio_lineedit->setText(settings.value("desktop_audio").toString());
    ui->disable_audio_checkbox->setChecked(settings.value("disable_audio_switch").toBool());
    ui->checkrate_slider->setValue(settings.value("checkrate").toInt(1000));
    ui->close_discord_checkbox->setChecked(settings.value("discord_action").toBool());
    ui->performance_powerplan_checkbox->setChecked(settings.value("powerplan_action").toBool());
    ui->gamemode_monitor_combobox->setCurrentIndex(settings.value("gamemode_monitor").toInt(0));
    ui->desktop_monitor_combobox->setCurrentIndex(settings.value("desktop_monitor").toInt(0));
    ui->disable_monitor_checkbox->setChecked(settings.value("disable_monitor_switch").toBool());

    // Call toggle functions based on settings
    toggleAudioSettings(!ui->disable_audio_checkbox->isChecked());
    toggleMonitorSettings(!ui->disable_monitor_checkbox->isChecked());
}

void BigPictureTV::saveSettings()
{
    qDebug() << "saving settings.";

    settings["gamemode_audio"] = ui->gamemode_audio_lineedit->text();
    settings["desktop_audio"] = ui->desktop_audio_lineedit->text();
    settings["disable_audio_switch"] = ui->disable_audio_checkbox->isChecked();
    settings["checkrate"] = ui->checkrate_slider->value();
    settings["discord_action"] = ui->close_discord_checkbox->isChecked();
    settings["powerplan_action"] = ui->performance_powerplan_checkbox->isChecked();
    settings["gamemode_monitor"] = ui->gamemode_monitor_combobox->currentIndex();
    settings["desktop_monitor"] = ui->desktop_monitor_combobox->currentIndex();
    settings["disable_monitor_switch"] = ui->disable_monitor_checkbox->isChecked();

    QFile file(settingsFile);
    if (file.open(QIODevice::WriteOnly)) {
        QJsonDocument doc(settings);
        file.write(doc.toJson(QJsonDocument::Indented));
        file.close();
    } else {
        qWarning() << "Failed to open settings file for writing.";
    }
}

void BigPictureTV::toggleAudioSettings(bool state)
{
    ui->audio_output_label->setEnabled(state);
    ui->gamemode_audio_label->setEnabled(state);
    ui->gamemode_audio_lineedit->setEnabled(state);
    ui->desktop_audio_label->setEnabled(state);
    ui->desktop_audio_lineedit->setEnabled(state);
}

void BigPictureTV::toggleMonitorSettings(bool state)
{
    ui->monitor_configuration_label->setEnabled(state);
    ui->gamemode_monitor_label->setEnabled(state);
    ui->gamemode_monitor_combobox->setEnabled(state);
    ui->desktop_monitor_label->setEnabled(state);
    ui->desktop_monitor_combobox->setEnabled(state);
}
